server/schedule/placement/region_rule_cache.go

```diff
@@ -0,0 +1,150 @@
+// 2. RegionFit is satisfied
+// RegionRuleFitCacheManager will invalid the cache for the region only when:
+// 1. region peer topology is changed
+// 2. region have down peers
+// 3. region leader is changed
+// 4. any involved rule is changed
+// 5. stores topology is changed
+// 6. any store label is changed
+// 7. any store state is changed
+type RegionRuleFitCacheManager struct {
+	mu     sync.RWMutex
+	caches map[uint64]*RegionRuleFitCache
+}
+
+// NewRegionRuleFitCacheManager returns RegionRuleFitCacheManager
+func NewRegionRuleFitCacheManager() *RegionRuleFitCacheManager {
+	return &RegionRuleFitCacheManager{
+		caches: map[uint64]*RegionRuleFitCache{},
+	}
+}
+
+// GetCacheRegionFit get RegionFit result by regionID
+func (manager *RegionRuleFitCacheManager) GetCacheRegionFit(regionID uint64) *RegionFit {
+	manager.mu.RLock()
+	defer manager.mu.RUnlock()
+	return manager.caches[regionID].bestFit
+}
+
+// Invalid invalid cache by regionID
+func (manager *RegionRuleFitCacheManager) Invalid(regionID uint64) {
+	manager.mu.Lock()
+	defer manager.mu.Unlock()
+	delete(manager.caches, regionID)
+}
+
+// InvalidAll invalids all cache
+func (manager *RegionRuleFitCacheManager) InvalidAll() {
+	manager.mu.Lock()
+	defer manager.mu.Unlock()
+	manager.caches = make(map[uint64]*RegionRuleFitCache)
+}
+
+// Check checks whether the region and rules are changed for the stored cache
+func (manager *RegionRuleFitCacheManager) Check(region *core.RegionInfo, rules []*Rule, stores []*core.StoreInfo) bool {
+	manager.mu.RLock()
+	defer manager.mu.RUnlock()
+	if cache, ok := manager.caches[region.GetID()]; ok && cache.bestFit != nil {
+		return cache.IsUnchanged(region, rules, stores)
+	}
+	return false
+}
+
+// SetCache stores RegionFit cache
+func (manager *RegionRuleFitCacheManager) SetCache(region *core.RegionInfo, rules []*Rule, fit *RegionFit, stores []*core.StoreInfo) {
+	manager.mu.Lock()
+	defer manager.mu.Unlock()
+	manager.caches[region.GetID()] = &RegionRuleFitCache{
+		region:  region,
+		rules:   rules,
+		bestFit: fit,
+		stores:  stores,
+	}
+}
+
+// RegionRuleFitCache stores regions RegionFit result and involving variables
+type RegionRuleFitCache struct {
+	bestFit *RegionFit
+	region  *core.RegionInfo
+	rules   []*Rule
+	stores  []*core.StoreInfo
+}
+
+// IsUnchanged checks whether the region and rules unchanged for the cache
+func (cache *RegionRuleFitCache) IsUnchanged(region *core.RegionInfo, rules []*Rule, stores []*core.StoreInfo) bool {
+	return cache.isRegionUnchanged(region) && rulesEqual(cache.rules, rules) && storeSEqual(cache.stores, stores)
+}
+
+func (cache *RegionRuleFitCache) isRegionUnchanged(region *core.RegionInfo) bool {
+	// we only cache region when it doesn't have down peers
+	if len(region.GetDownPeers()) > 0 || region.GetLeader() == nil {
+		return false
+	}
+	return region.GetLeader().StoreId == cache.region.GetLeader().StoreId && peersEqual(cache.region, region)
+}
+
+func peersEqual(a, b *core.RegionInfo) bool {
```

Comment:
can we replace it by checking region's confVersion?

Reply:
updated.